tests/quantum_locks_tests.cpp

```diff
@@ -351,6 +354,100 @@ TEST(ReadWriteSpinLock, Guards)
     EXPECT_FALSE(lock.isLocked());
 }
 
+TEST(ReadWriteSpinLock, UpgradeLock)
+{
+    ReadWriteSpinLock lock;
+    lock.lockRead();
+    lock.lockRead();
+    lock.lockRead();
+    EXPECT_TRUE(lock.isLocked());
+    EXPECT_TRUE(lock.isReadLocked());
+    EXPECT_FALSE(lock.isWriteLocked());
+    EXPECT_EQ(3, lock.numReaders());
+    EXPECT_EQ(0, lock.numPendingWriters());
+    std::thread t([&lock]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
This part of the test looks very timing-specific and thus fragile. Let's find a more robust method of coordination; perhaps check out some of the tests in my PR in the private quantum?
My concern is that this basically assumes that if the thread sleeps 100ms, the upgradeToWrite() call will have started. That should be true, but if for some reason the system is slow, the test fails.
Perhaps we could just have the thread do
while (lock.numPendingWriters() == 0) {} instead of the sleep